graphbolt/src/csc_sampling_graph.cc

```diff
@@ -131,16 +131,144 @@ c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::InSubgraph(
           : torch::nullopt);
 }
 
+NumPickFn GetNumPickFnWithoutEtype(
+    int64_t fanout, bool replace,
+    const torch::optional<torch::Tensor>& probs_or_mask) {
+  if (!probs_or_mask.has_value()) {
+    return [fanout, replace](int64_t offset, int64_t num_neighbors) {
+      int64_t max_count = fanout == -1 ? num_neighbors : fanout;
+      if (replace) {
+        return num_neighbors == 0 ? 0 : max_count;
+      } else {
+        return std::min(num_neighbors, max_count);
+      }
+    };
+  } else {
+    return [fanout, replace, probs_or_mask](
+               int64_t offset, int64_t num_neighbors) {
+      int64_t max_count = fanout == -1 ? num_neighbors : fanout;
+      int64_t cur_count = 0;
+      AT_DISPATCH_ALL_TYPES(
+          probs_or_mask.value().scalar_type(), "NumPickFn", ([&] {
+            const scalar_t* probs_data =
+                probs_or_mask.value().data_ptr<scalar_t>();
+            for (int64_t i = offset; i < offset + num_neighbors; ++i) {
+              if (probs_data[i] != 0) ++cur_count;
+            }
+          }));
+      if (replace) {
+        return cur_count == 0 ? 0 : max_count;
+      } else {
+        return std::min(cur_count, max_count);
+      }
+    };
+  }
+}
+
+NumPickFn GetNumPickFnWithEtype(
+    const std::vector<int64_t>& fanouts, bool replace,
+    const torch::optional<torch::Tensor>& type_per_edge,
```

Comment:
When call this func, `type_per_edge` must has value, thus use `Tenosr `directly without `optional`.